File: Java/problem1382.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private List<Integer> sortedLst;
    
    private void inorderTraversal(TreeNode node) {
        if (node == null) {
            return;
        }
        inorderTraversal(node.left);
        sortedLst.add(node.val);
        inorderTraversal(node.right);
    }
    
    private TreeNode buildTree(int leftIdx, int rightIdx) {
        if (leftIdx > rightIdx) {
            return null;
        }
        if (leftIdx == rightIdx) {
            return new TreeNode(sortedLst.get(leftIdx));
        }
        int middleIdx = leftIdx + (rightIdx - leftIdx) / 2;
        TreeNode root = new TreeNode(sortedLst.get(middleIdx));
        root.left = buildTree(leftIdx, middleIdx - 1);
        root.right = buildTree(middleIdx + 1, rightIdx);
        return root;
    }
    
    public TreeNode balanceBST(TreeNode root) {
        sortedLst = new ArrayList<>();
        inorderTraversal(root);
        return buildTree(0, sortedLst.size() - 1);
    }
}

File: python/problem1382.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        sortedLst = []
        def inorderTraversal(node):         #中序遍历，得到一个排好序的节点值列表
            if node is None:
                return
            inorderTraversal(node.left)
            sortedLst.append(node.val)
            inorderTraversal(node.right)

        def bulidTree(left, right):         #以列表中间节点为根节点，左边节点为左子树，右边节点为右子树
            if left > right:
                return None
            if left == right:
                return TreeNode(sortedLst[left])
            middldIdx = left + (right - left) // 2
            node = TreeNode(sortedLst[middldIdx])
            node.left = bulidTree(left, middldIdx - 1)
            node.right = bulidTree(middldIdx + 1, right)
            return node
        
        inorderTraversal(root)
        return bulidTree(0, len(sortedLst) - 1)
